### ISSUE TITLE
Add ability to MethodDescriptor to return classes of request and response.

### ISSUE THREAD

**Issue by knok16** [grpc/grpc-java#1814]
Hi,

`MethodDescriptor` is parametrized by `<RequestT, ResponseT>` but programmatically it's not possible to get this classes. What do you think about adding 2 getters to `MethodDescriptor` and populating it in generated classes:

`Class<RequestT> getRequestClass()` / `RequestT getRequestDefaultInstance() //returns new instance each call`
`Class<ResponseT> getResponseClass()` / `ResponseT getResponseDefaultInstance() //returns new instance each call`.

Thanks.

**Comment by buchgr** [comment]
Hi @knok16 !

Why do you need this? What's your use case?

I believe adding `getRequestClass()` would require quite a bit of hacking with the reflection API, due to type erasure. Also, what would `getRequestDefaultInstance` return? Note that `RequestT` and `ResponseT` can be of ANY type, and don't need to be protobuf generated classes.

**Comment by buchgr** [comment]
@knok16 Actually, if you really need this you can implement this yourself with some hacking.

Note the two methods `(ResponseT|RequestT) MethodDescriptor.parse(Response|Request)(InputStream)`. These just delegate to the request and response `(ResponseT|RequestT) Marshaller.parse(InputStream) ` which you provide yourself in the `MethodDescriptor` constructor. So you have full control over what to return there. For example, in your own project you could re-define the API so that if `null` is passed to the `parse...(InputStream)` method, that it returns a `getDefaultInstance()` object. You could then put this functionality into a utility class i.e. `MethodDescriptors.getRequestClass(MethodDescriptor)`.

**Comment by knok16** [comment]
I was going to implement http endpoints (with use of [ProtoUtils](https://github.com/grpc/grpc-java/blob/master/protobuf/src/main/java/io/grpc/protobuf/ProtoUtils.java)) and also it can be used for work with MethodDescriptor in generic way:

e.g. 
```
final ServerServiceDefinition serverServiceDefinition = ...;
for (final ServerMethodDefinition<?, ?> method : serverServiceDefinition.getMethods()) {
    final MethodDescriptor<?, ?> methodDescriptor = method.getMethodDescriptor();
    // There is no any clue what class of request/response
}
```

Yes, I can construct new `MethodDescriptor` with any marshallers - but it should be done by hand for each method, and if methods will be added - it should be done again and again.

Also one point for `getRequestClass/getResponseClass` methods is that actually `MethodDescriptor` is already parametrized and have this info - we just not able to get it - due to java internal organization (for example C# allows get perameter's classes throught reflection).

> Note that RequestT and ResponseT can be of ANY type, and don't need to be protobuf generated classes.

Agree - then `getRequestDefaultInstance` is definitly not an option.

I thought that request/responses classes will added to `MethodDescriptor.create(...)` to populate in `MethodDescriptor` and will be provided in generated classes:

```
public MethodDescriptor<echo.EchoOuterClass.StringRequest, echo.EchoOuterClass.StringResponse> METHOD_ECHO_STRING =
  MethodDescriptor.create(
      MethodDescriptor.MethodType.UNARY,
      generateFullMethodName("echo.Echo", "echoString"),
      io.grpc.protobuf.ProtoUtils.marshaller(echo.EchoOuterClass.StringRequest.getDefaultInstance()),
      io.grpc.protobuf.ProtoUtils.marshaller(echo.EchoOuterClass.StringResponse.getDefaultInstance()),
      EchoOuterClass.StringRequest.class,
      EchoOuterClass.StringResponse.class
      );
```

**Comment by buchgr** [comment]
> I was going to implement http endpoints (with use of ProtoUtils)

Can you please be a bit more specific about why exactly you need that? Maybe some sample code? There might be other ways to achieve your goal.

> Also one point for getRequestClass/getResponseClass methods is that actually MethodDescriptor is already parametrized and have this info - we just not able to get it - due to java internal organization (for example C# allows get perameter's classes throught reflection).

I think we might be able to do it using  https://docs.oracle.com/javase/7/docs/api/java/lang/reflect/ParameterizedType.html.

**Comment by knok16** [comment]
```
final ServerServiceDefinition serverServiceDefinition = ...
final Map<String, HttpHandler> endpoints = serverServiceDefinition.getMethods().stream()
        .map(ServerMethodDefinition::getMethodDescriptor)
        .collect(Collectors.toMap(m -> "/" + m.getFullMethodName(),
                methodDescriptor -> {
                    // Here I'm not able to get appropriate marshaller in generic way because I don't know class of the Response
                    final MethodDescriptor.Marshaller<Message> marshaller = getMarshaller(methodDescriptor);
                    return httpExchange -> {
                        final ClientCall<Message, Message> call = channel.newCall((MethodDescriptor<Message, Message>) methodDescriptor, CallOptions.DEFAULT);
                        call.start(
                                getResponseListener(httpExchange),
                                getHeaders(httpExchange.getRequestHeaders())
                        );
                        call.sendMessage(marshaller.parse(httpExchange.getRequestBody()));
                        call.halfClose();
                        call.request(1);
                    };
                }
        ));
```

**Comment by IschenkoArtem** [comment]
Any news on that? 
I'm also interested in seeing this last piece of Type Info in MethodDescriptor, without it it's hard to build instrumentation on top of it.

**Comment by ejona86** [comment]
I think we'll support this by allowing getting the marshaller. The Protobuf marshaller than could implement an interface for getting the default instance.

**Comment by ejona86** [comment]
This has been fixed since #2242
